src/transaction.ts

```diff
@@ -180,6 +184,75 @@ function ensureBoxReference(input: unknown): TransactionBoxReference {
   throw new Error(`Not a box reference: ${input}`);
 }
 
+export interface TransactionHoldingReference {
+  readonly assetIndex: bigint;
+  readonly address: Address;
+}
+
+function ensureHoldingReference(input: unknown): TransactionHoldingReference {
+  if (input != null && typeof input === 'object') {
+    const { assetIndex, address } = input as TransactionHoldingReference;
+    return {
+      assetIndex: utils.ensureUint64(assetIndex),
+      address: ensureAddress(address),
+    };
+  }
+  throw new Error(`Not a holding reference: ${input}`);
+}
+
+export interface TransactionLocalsReference {
+  readonly appIndex: bigint;
+  readonly address: Address;
+}
+
+function ensureLocalsReference(input: unknown): TransactionLocalsReference {
+  if (input != null && typeof input === 'object') {
+    const { appIndex, address } = input as TransactionLocalsReference;
+    return {
+      appIndex: utils.ensureUint64(appIndex),
+      address: ensureAddress(address),
+    };
+  }
+  throw new Error(`Not a locals reference: ${input}`);
+}
+
+export interface TransactionResourceReference {
+  readonly address?: Readonly<Address>;
+  readonly appIndex?: Readonly<bigint>;
+  readonly assetIndex?: Readonly<bigint>;
+  readonly holding?: Readonly<TransactionHoldingReference>;
+  readonly locals?: Readonly<TransactionLocalsReference>;
+  readonly box?: Readonly<TransactionBoxReference>;
+}
+
+function ensureResourceReference(input: unknown): TransactionResourceReference {
+  if (input != null && typeof input === 'object') {
+    const { address, appIndex, assetIndex, holding, locals, box } =
+      input as TransactionResourceReference;
+    let result = {};
+    if (address !== undefined) {
+      result = { ...result, address: ensureAddress(address) };
+    }
+    if (appIndex !== undefined) {
+      result = { ...result, appIndex: utils.ensureUint64(appIndex) };
+    }
+    if (assetIndex !== undefined) {
+      result = { ...result, assetIndex: utils.ensureUint64(assetIndex) };
+    }
+    if (holding !== undefined) {
+      result = { ...result, holding: ensureHoldingReference(holding) };
+    }
+    if (locals !== undefined) {
+      result = { ...result, locals: ensureLocalsReference(locals) };
+    }
+    if (box !== undefined) {
+      result = { ...result, box: ensureBoxReference(box) };
+    }
+    return result;
```

Comment:
Since only one of these can be true, instead of accumulating into `result`, can you just return it?